### PR TITLE
Better operator dimension parameterisation

### LINKED ISSUE #1722: Better operator dimension parameterisation
**Is your feature request related to a problem? Please describe.**
The dimension argument of qutip operators are quite inconsistent in the **naming**, **type** and **description**. Here are a few examples picked from the [functions doc page](https://qutip.org/docs/latest/apidoc/functions.html):
- `basis` > `dimensions: int or list of ints` _Number of Fock states in Hilbert space. If a list, then the resultant object will be a tensor product over spaces with those dimensions._
- `enr_fock` > `dims: list` _A list of the dimensions of each subsystem of a composite quantum system._
- `enr_state_dictionaries` > `dims: list` _A list with the number of states in each sub-system._
- `phase_basis` > `N: int` _Number of basis vectors in Hilbert space._
- `projection` > `N: int` _Number of basis states in Hilbert space._
- `momentum` > `N: int` _Number of Fock states in Hilbert space._
- `num` > `N: int` _The dimension of the Hilbert space._
- `qeye` > `dimensions: (int) or (list of int) or (list of list of int)` _Dimension of Hilbert space. If provided as a list of ints, then the dimension is the product over this list, but the dims property of the new Qobj are set to this list. This can produce either oper or super depending on the passed dimensions._

Is there a reason for this apparent disorder? I'd be happy to propose a PR.

**Describe the solution you'd like**
Harmonise the naming, type and description of the operators dimension argument.

**Describe alternatives you've considered**
Even if not all operators are meant to work on composite Hilbert spaces, harmonising the naming and docstring would already be helpful.

**Additional context**
Cc @francois-marie for indicating the issue.

### PR DESCRIPTION
**Pull request by PriyanshuPansari**
Removed the inconsistency in the dimension argument of Qutip operators in the naming, type, and description.
Proposed generalization - use N to denote the dimension of Hilbert space if it's int or denote a list with the number of states in each sub-system if it's a list.

fixes #1722

### CODE REVIEW

File: qutip/core/states.py [hunk 1]
```diff
@@ -42,22 +43,23 @@ def _promote_to_zero_list(arg, length):
     raise TypeError("Dimensions must be an integer or list of integers.")
 
 
-def basis(dimensions, n=None, offset=None, *, dtype=_data.Dense):
+def basis(N, n=None, offset=None, *, dtype=_data.Dense):
```
Comment (Ericgig) [inline qutip/core/states.py:46]: Here `dimensions` is a better name than `N`. It's more descriptive and I would expect `N` to only accept an integer.
`list of int` are the way we represent the dimension of the state.

Comment (PriyanshuPansari) [reply qutip/core/states.py:46]: @Ericgig Ok so for functions that take int or list of int I should use dimensions and if it takes only int then N or should I change all to dimensions?

**Review by Ericgig — COMMENTED** [review]
There are multiple input types that should be treated differently:
- dimensions : int or list of int
  Operators and states well defined on complex (tensor) space.
- N : int
  Operators and states well defined on simple space.
- dims : list of int
  enr_... functions

Changing all the name to be the same, but keeping different input type will cause more confusion. 

Here the tasks I see:
- Ensure one name for each input type.
- Have the documentation match.
- Do the same for `qutip/core/operators.py`
- Review the functions and confirm that the right input type is used.

File: qutip/core/states.py [hunk 2]
```diff
@@ -73,15 +75,15 @@ def basis(dimensions, n=None, offset=None, *, dtype=_data.Dense):
     Examples
     --------
     >>> basis(5,2) # doctest: +SKIP
-    Quantum object: dims = [[5], [1]], shape = (5, 1), type = ket
+    Quantum object: N = [[5], [1]], shape = (5, 1), type = ket
```
[the remark this answers is not in the document]
Comment (PriyanshuPansari) [reply qutip/core/states.py:78]: Will make the changes

File: qutip/core/states.py [hunk 3]
```diff
@@ -974,35 +976,34 @@ def enr_fock(dims, excitations, state, *, dtype=_data.Dense):
...
         A Qobj instance that represent a Fock state in the exication-number-
-        restricted state space defined by `dims` and `exciations`.
+        restricted state space defined by `N` and `exciations`.
 
     """
-    nstates, state2idx, _ = enr_state_dictionaries(dims, excitations)
+    nstates, state2idx, _ = enr_state_dictionaries(N, excitations)
     try:
-        data =_data.one_element[dtype]((nstates, 1),
+        data = data.one_element[dtype]((nstates, 1),
```
[the remark this answers is not in the document]
Comment (PriyanshuPansari) [reply qutip/core/states.py:984]: Sorry, changed it by mistake.

**Pull request closed without merging**